batchflow/models/torch/modules.py

```diff
@@ -0,0 +1,228 @@
+        Kernel sizes in pyramid block convolutions
+    layout: str
+        Layout for convolution layers.
+    downsample_layout: str
+        Layout for downsampling layers. Default is 'p'
+    upsample_layout: str
+        Layout for upsampling layers. Default is 't
+    factor: int
+        Scaling factor for upsampling layers. Default is 2
+    bottleneck : bool, int
+        If True, then add a 1x1 convolutions before and after pyramid block with that factor of filters reduction.
+        If False, then bottleneck is not used. Default is False.
+    use_dilation: bool
+        If True, then the convolutions with bigger kernels in the pyramid block are replaced by top one
+        with corresponding dilation_rate, i.e. 5x5 -> 3x3 with dilation=2, 7x7 -> 3x3 with dilation=3.
+        If False, the dilated convolutions are not used. Default is False.
+    """
+    def __init__(self, inputs=None, pyramid_kernel_size=(7, 5, 3), bottleneck=False, layout='cna',
+                 downsample_layout='p', upsample_layout='t', factor=2, use_dilation=False, **kwargs):
+        super().__init__()
+        depth = len(pyramid_kernel_size)
+        spatial_shape = get_shape(inputs)[2:]
+        num_dims = get_num_dims(inputs)
+        self.attention = ConvBlock(inputs=inputs, layout='V >' + layout + 'b', kernel_size=1,
+                                   filters='same', shape=spatial_shape, dim=num_dims, **kwargs)
+
+        enc_layout = ('B' + downsample_layout + layout) * depth # B pcna B pcna B pcna
+        emb_layout = layout + upsample_layout # cnat
+        combine_layout = ('+' + upsample_layout) * (depth - 1) + '*' # +t+t*
+        main_layout = enc_layout + emb_layout + combine_layout # B pcna B pcna B pcna cnat +t+t*
+        main_strides = [1] * (depth + 1) + [factor] * depth # [1, 1, 1, 1, 2, 2, 2]
+
+        # list with args for BaseBlocks of each branch
+        branches = [dict(layout=layout, filters='same', kernel_size=1)] # the mid branch from inputs tensor directly
+
+        if use_dilation:
+            base_kernel_size = pyramid_kernel_size[-1] # 3
+            main_kernel_size = [base_kernel_size] * (depth + 1) + [factor] * (depth) # 3 3 3 3 2 2 2
+            # infering coresponding dilation for every kernel_size in pyramid block
+            pyramid_dilation = [round((rf - base_kernel_size) / (base_kernel_size - 1)) + 1
+                                for rf in pyramid_kernel_size] # 3 2 1
+            main_dilation = pyramid_dilation + [pyramid_dilation[-1]] + [1] * depth # 3 2 1 1 1 1 1
+            for d in pyramid_dilation[:-1]:
+                args = dict(layout=layout, kernel_size=base_kernel_size, dilation_rate=d, filters='same')
+                branches.append(args)
+        else:
+            main_kernel_size = pyramid_kernel_size + [pyramid_kernel_size[-1]] + [factor] * (depth) # [7 5 3 3 2 2 2]
+            main_dilation = 1
+            for kernel_size in pyramid_kernel_size[:-1]:
```

Comment:
That loop is almost the same in two branches of `if-else` and can be easily generalized to one